[Feature Contribution] Scafacos

rhalver opened the pull request:
## Purpose

Fix for the ScaFaCoS library[1], a library for the computation of electro-statics, providing several solvers to choose from, e.g. FMM or P3M. The library was developed in a project funded by the
German Ministry of Science and Education from 2009-2012.

## Author(s)

Rene Halver, Juelich Supercomputing Centre, Forschungszentrum Juelich, Germany

## Backward Compatibility

No changes in backward compatibility are expected.

## Implementation Notes

No other features are influenced by this fix. Within the fix the Coulomb field and potentials
are computed and resulting from them the forces and energy are calculated.Some
changes might be required to completely adhere to the LAMMPS environment.

## Post Submission Checklist

_Please check the fields below as they are completed_
- [ ] The feature or features in this pull request is complete
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [X] One or more example input decks are included
- [X] The source code follows the LAMMPS formatting guidelines

## Further Information, Files, and Links

[1] www.scafacos.de

stanmoore1 commented:
For convenience, can the scafacos library be bundled with LAMMPS in /lib, or is there a reason why not, like it is too large or the licenses are incompatible?

akohlmey commented on src/USER-SCAFACOS/fix_scafacos.h:
```diff
@@ -0,0 +1,164 @@
+#endif
+#endif
+
+/* ERROR/WARNING messages:
```
This part of the header lists errors and warnings from fix latte. This should instead list the corresponding errors and warnings in the fix scafacos code and the references to fix latte removed.

akohlmey commented on src/USER-SCAFACOS/fix_scafacos.cpp:
```diff
@@ -0,0 +1,454 @@
+{
+  if (result) 
+  {
+    printf("ScaFaCoS Error: Caught error on task %d.\n", comm_rank);
```
Please don't use `printf()`, but `fprintf(screen,...)` and `fprintf(logfile, ...)` provided either FILE pointer is non-NULL. Please note, that calling error->all() requires, that this function is called from *all* MPI ranks on the communicator `world`, otherwise you need to call error->one(). If however `check_results()` *is* called from all MPI ranks on `world`, then only MPI rank 0 should be producing output from `fprintf()` or else the screen would have all messages replicated by the number of MPI ranks in use.

stanmoore1 commented:
@rhalver does this work with the original `gmake` build of LAMMPS or only with the CMake? I tried `gmake` but the `Makefile.lammps` file in /lib/scafacos is missing and the build failed.

rhalver commented on src/USER-SCAFACOS/fix_scafacos.cpp:
```diff
@@ -439,15 +439,15 @@ bool FixScafacos::check_result(FCSResult result, int comm_rank)
   if (result) 
   {
-    printf("ScaFaCoS Error: Caught error on task %d.\n", comm_rank);
+    fprintf(stdout,"ScaFaCoS Error: Caught error on task %d.\n", comm_rank);
```
Ok I think the easiest way is to remove the fprintf statement and just pass the internal error message to the error -> one() call. As it is possible that a subset of processes (>1) may encounter an error, is error -> one() still the correct call?

stanmoore1 replied:
>As it is possible that a subset of processes (>1) may encounter an error, is error -> one() still the correct call?

Yes this is fine. You can only use `error->all` if you can guarantee that all processes will reach it, otherwise it will deadlock.

junghans commented on cmake/CMakeLists.txt:
```diff
@@ -185,11 +185,11 @@ pkg_depends(CORESHELL KSPACE)
 endif()
 
-if(PKG_MEAM OR PKG_USER-H5MD OR PKG_USER-QMMM)
+if(PKG_MEAM OR PKG_USER-H5MD OR PKG_USER-QMMM OR PKG_USER-SCAFACOS)
```
I don't think that is needed! Can you try again if it works without it!

junghans commented on cmake/CMakeLists.txt:
```diff
@@ -348,6 +348,13 @@ if(PKG_LATTE)
 
+if(PKG_USER-SCAFACOS)
+    FIND_PACKAGE(PkgConfig)
+    PKG_CHECK_MODULES(SCAFACOS scafacos)
```
Need to add `REQUIRED` here.

And we need to add a `DOWNLOAD_SCAFACOS` option.

sjplimp commented:
@rhalver I made some edits to the 2 KSpace doc pages.  Added a few RENE comments/Qs
that you can address

junghans commented on cmake/CMakeLists.txt:
```diff
@@ -349,6 +351,58 @@ if(PKG_LATTE)
+#libfcs.a           libfcs_direct.a   libfcs_ewald.la   libfcs_fftw3_mpi.a  libfcs_fmm.la        libfcs.la           libfcs_p2nfft.a  libfcs_pfft.la    libfcs_resort.a
+  else()
+    FIND_PACKAGE(PkgConfig REQUIRED)
+    PKG_CHECK_MODULES(SCAFACOS scafacos)
```
A `REQUIRED` keyword is missing here!

junghans commented:
I tested this again and it works for me now.

sjplimp commented:
@rhalver Rene - I added the PBC enforcement before the call to scafacos, only
changed scafacos.cpp/h files

sjplimp commented:
@rhalver I did 2 things in this push: (a) updated with all the current master files, so I could add Scafacos info to the new doc pages, (b) made one change to your latest scafacos,cpp for the call to domain->remap().  See my RENE comments in the code for this.

sjplimp commented:
@akohlmey @junghans @rbberger we think (me and @rhalver) this is close to ready to be
merged.  Do you want to take a look and make any needed changes.  Note that this 
adds a lib/scafacos and src/USER-SCAFACOS.  I think Christoph has already added the CMake
support for building the ScaFaCoS lib.

junghans commented on cmake/CMakeLists.txt:
```diff
@@ -139,6 +139,34 @@ set(LAMMPS_LINK_LIBS)
+  enable_language(C)
+endif()
+
+# do MPI detection after language activation, if MPI for these language is required
```
This is the reason why the block had to be moved up.

junghans commented on cmake/CMakeLists.txt:
```diff
@@ -400,6 +402,57 @@ if(PKG_LATTE)
+    list(APPEND LAMMPS_LINK_LIBS ${SCAFACOS_BUILD_DIR}/lib/libfcs_pnfft.a)
+    list(APPEND LAMMPS_LINK_LIBS ${SCAFACOS_BUILD_DIR}/lib/libfcs_pfft.a)
+    list(APPEND LAMMPS_LINK_LIBS ${SCAFACOS_BUILD_DIR}/lib/libfcs_fftw3_mpi.a)
+    list(APPEND LAMMPS_LINK_LIBS ${SCAFACOS_BUILD_DIR}/lib/libfcs_fftw3.a)
```
@rbberger do have a better idea than this explicit listing?

rbberger replied:
@junghans not really, although hardcoding some of those (libm) seems wrong. We have that at some other place. Same goes for FFTW3, what is the benefit of having the internal one? Also, this breaks with `-D BUILD_LIB=on -D BUILD_SHARED_LIBS=on` due to missing `-fPIC`
```
/usr/bin/ld: scafacos_build-prefix/lib/libfcs_fmm.a(libfcs_fmm_back_idx_la-dash.o): relocation R_X86_64_PC32 against symbol `fcs_back_idx_sl_default_context' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/lammps.dir/build.make:6050: liblammps.so.0] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/lammps.dir/all] Error 2
```

junghans replied:
@rbberger, I fixed the `PIC` issue.

junghans replied:
@rbberger, there is a `--with-fftw3-libdir=DIR` and `--with-fftw3-includedir=DIR` option, so we could detect FFTW3 in CMake and hand this over to Scafacos if found or inject `--with-internal-fftw` if FFT3 isn't found.

akohlmey commented on src/USER-SCAFACOS/scafacos.h:
```diff
@@ -0,0 +1,74 @@
+#define LMP_SCAFACOS_H
+
+#include "kspace.h"
+#include "fcs.h"
```
this statement must not be here. as commented before, we have to keep our namespace clean, which means all variables/pointers that require types from a header could be encapsulated into a pointer to a struct (for which we can put a forward declaration into the header) or should be cast to void pointers or stored in compatible regular types and then the header specific types can be used and the header included in the implementation, i.e. the .cpp file only. we had a bunch of issues because of namespace pollution already and it took some effort to get this cleaned up in almost all places where this was possible without having to rewrite large parts of the code (some parts like USER-REAXC, this was only done partially).

akohlmey commented:
cleaner and simpler (IMO) would be to put a

struct fcsdata *fcs_handle;

into the class definition and then

define its structure inside the .cpp file and move all specific data types
that require definitions from the fcs.h header into this struct instead of
the LAMMPS kspace style class.
you can then access them with either handle->varname instead of varname, or
- if performance is an issue - cache it as mytype varname = handle->varname;
LAMMPS does this with its own data structures (e.g. Atom class) all over
the place. this results in much better type safety and provides the same
namespace hygiene than casting to void pointers or compatible types and
back.

On Wed, Aug 22, 2018 at 5:24 PM Steve Plimpton <notifications@github.com>
wrote:

> @rhalver <https://github.com/rhalver> - the issue Axel flagged is to
> remove the fcs.h file from scafacos.h, which also means
> all the fcs and FCS declarations. I think fcs (handle) could just be void
> *fcs_handle, and converted back
> to an FCS pointer when used in scafacos.cpp. Result could just be dropped
> from the *.h since you
> could declare it whenever used in *.cpp. The fcs int/float could be
> regular int and double and the ScaF methods that use them could copy values
> in/out of them into locally declared fcs_int/float.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/989#issuecomment-415185653>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARp4x9vFlSqs1s5Eq1vT6MWmSBfP7uCks5uTcwYgaJpZM4VIWf0>
> .
>


-- 
Dr. Axel Kohlmeyer  akohlmey@gmail.com  http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste. Italy.

rhalver commented:
To make sure I don't missunderstand anything again, I should completely remove
the inclusion from the header, i.e. scafacos.h, but it would be alright to include
the fcs.h in the scafacos.cpp?

While I agree that that is a cleaner solution to define proxy structs, I am not
sure if is simpler in this case, as the FCS data type which is used here is only a
pointer to a struct used as a data container for internal data. Within
scafacos.cpp I never directly manipulate data within it, but the FCS variable is just
passed to the corresponding ScaFaCoS routines, which process its contents.
Using a void* to define it in the class and cast it in the function calls would
mirror what we do in our Fortran interface, as the FCSHandle is represented
with a (renamed) c_ptr type there.

What seems to be more of a hassle to me is the check_result function, that
has a FCSResult argument. I could rewrite it in a similar way, so that a void*
is expected and inside the function cast it to the correct FCSResult data type.

Rene Halver
Institute for Advanced Simulation - Juelich Supercomputing Centre
Telefon 02461 61-6424



------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------
Forschungszentrum Juelich GmbH
52425 Juelich
Sitz der Gesellschaft: Juelich
Eingetragen im Handelsregister des Amtsgerichts Dueren Nr. HR B 3498
Vorsitzender des Aufsichtsrats: MinDir Dr. Karl Eugen Huthmacher
Geschaeftsfuehrung: Prof. Dr.-Ing. Wolfgang Marquardt (Vorsitzender),
Karsten Beneke (stellv. Vorsitzender), Prof. Dr.-Ing. Harald Bolt,
Prof. Dr. Sebastian M. Schmidt
------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------

akohlmey commented:
On Wed, Aug 22, 2018 at 7:11 PM rhalver <notifications@github.com> wrote:

> To make sure I don't missunderstand anything again, I should completely
> remove
> the inclusion from the header, i.e. scafacos.h, but it would be alright to
> include
> the fcs.h in the scafacos.cpp?
>

how you implement this is up to you. what matters to the maintenance of
LAMMPS is that fcs.h is removed from scafacos.h.

there were multiple variables in the class definition with fcs_int and
fcs_float type, if you don't care to use int and double for those, then
indeed it would be simpler to cast the handle to void *. i was assuming
that you chose fcs_int and fcs_float for a reason and wanted to preserve
that. this is why i suggested to use a struct and bundle all of those
elements and hide them from the class definition.

axel.



>
> While I agree that that is a cleaner solution to define proxy structs, I
> am not
> sure if is simpler in this case, as the FCS data type which is used here
> is only a
> pointer to a struct used as a data container for internal data. Within
> scafacos.cpp I never directly manipulate data within it, but the FCS
> variable is just
> passed to the corresponding ScaFaCoS routines, which process its contents.
> Using a void* to define it in the class and cast it in the function calls
> would
> mirror what we do in our Fortran interface, as the FCSHandle is represented
> with a (renamed) c_ptr type there.
>
> What seems to be more of a hassle to me is the check_result function, that
> has a FCSResult argument. I could rewrite it in a similar way, so that a
> void*
> is expected and inside the function cast it to the correct FCSResult data
> type.
>
> Rene Halver
> Institute for Advanced Simulation - Juelich Supercomputing Centre
> Telefon 02461 61-6424
>
>
>
>
> ------------------------------------------------------------------------------------------------
>
> ------------------------------------------------------------------------------------------------
> Forschungszentrum Juelich GmbH
> 52425 Juelich
> Sitz der Gesellschaft: Juelich
> Eingetragen im Handelsregister des Amtsgerichts Dueren Nr. HR B 3498
> Vorsitzender des Aufsichtsrats: MinDir Dr. Karl Eugen Huthmacher
> Geschaeftsfuehrung: Prof. Dr.-Ing. Wolfgang Marquardt (Vorsitzender),
> Karsten Beneke (stellv. Vorsitzender), Prof. Dr.-Ing. Harald Bolt,
> Prof. Dr. Sebastian M. Schmidt
>
> ------------------------------------------------------------------------------------------------
>
> ------------------------------------------------------------------------------------------------
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lammps/lammps/pull/989#issuecomment-415216959>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARp40mMPQfg4dIuEu1ZCt1AFOiu5Tqeks5uTeUdgaJpZM4VIWf0>
> .
>


-- 
Dr. Axel Kohlmeyer  akohlmey@gmail.com  http://goo.gl/1wk0
College of Science & Technology, Temple University, Philadelphia PA, USA
International Centre for Theoretical Physics, Trieste. Italy.

rhalver commented:
With the last commit I removed fcs.h from scafacos.h and replaced the custom structures
with void*. Also I replaced the ScaFaCoS data types with int and double, which should be
the most common choices, but in principle it is possible to use other data types.. Is it possible to
run LAMMPS purely with float as it then make sense to provide the possibility
to run ScaFaCoS with float as well (with the use of fcs_float set to float)?

akohlmey commented on src/USER-SCAFACOS/scafacos.cpp:
```diff
@@ -0,0 +1,485 @@
+  qqrd2e = force->qqrd2e;
+
+  if (!initialized) {
+    printf("DEBUG: %p\n",&fcs);
```
please remove debug statements like these.
